src/Pusher.php

```diff
@@ -1073,4 +1006,153 @@ public function ensure_valid_signature(array $headers, string $body): void
     {
         $this->verifySignature($headers, $body);
     }
+
+    /**
+     * Returns an event represented by an associative array to be used in creating events and batch_events requests
+     *
+     * @param array|string $channels A channel name or an array of channel names to publish the event on.
+     * @param string $event
+     * @param mixed $data Event data
+     * @param array $params [optional]
+     * @param bool $already_encoded [optional]
+     *
+     * @throws PusherException
+     *
+     * @return array Event associative array
+     */
+    private function make_event(array $channels, string $event, $data, array $params = [], ?string $info = null, bool $already_encoded = false): array
+    {
+      $has_encrypted_channel = false;
+        foreach ($channels as $chan) {
+            if (PusherCrypto::is_encrypted_channel($chan)) {
+                $has_encrypted_channel = true;
+                break;
+            }
+        }
+
+        if ($has_encrypted_channel) {
+            if (count($channels) > 1) {
+                // For rationale, see limitations of end-to-end encryption in the README
+                throw new PusherException('You cannot trigger to multiple channels when using encrypted channels');
+            } else {
+                try {
+                    $data_encoded = $this->crypto->encrypt_payload(
+                        $channels[0],
+                        $already_encoded ? $data : json_encode($data, JSON_THROW_ON_ERROR)
+                    );
+                } catch (\JsonException $e) {
+                    throw new PusherException('Data encoding error.');
+                }
+            }
+        } else {
+            try {
+                $data_encoded = $already_encoded ? $data : json_encode($data, JSON_THROW_ON_ERROR);
+            } catch (\JsonException $e) {
+                throw new PusherException('Data encoding error.');
+            }
+        }
+
+        // json_encode might return false on failure
+        if (!$data_encoded) {
```

Comment:
This is copying from the previous implementation. So I'll leave it as is in this PR, but it could be done in a separate PR. There are some other long functions in this file.